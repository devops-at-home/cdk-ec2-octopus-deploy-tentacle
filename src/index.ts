import { ResourceProps } from 'aws-cdk-lib';
import { Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { CfnDocument } from 'aws-cdk-lib/aws-ssm';
import { Construct } from 'constructs';

interface OctopusDeployDocumentProps extends ResourceProps {
    secretArn: string;
}

export class OctopusDeployDocument extends Construct {
    constructor(scope: Construct, id: string, props: OctopusDeployDocumentProps) {
        super(scope, id);

        const {} = props;

        const { roleName } = new Role(this, 'Role', {
            assumedBy: new ServicePrincipal('ssm.amazon.com'),
            inlinePolicies: {},
        });

        new CfnDocument(this, 'CfnDocument', {
            content: getContent(roleName),
            documentType: 'Automation',
            name: 'ConfigureOctopusDeploy',
            targetType: '/AWS::EC2::Instance',
        });
    }
}

const getContent = (assumeRole: string) => {
    // From: https://github.com/linuxacademy/Hands-On-with-AWS-Systems-Manager/blob/master/Automation-Change-Mgmt/ssm-automation.json
    return {
        description:
            "Automation for creating an instance, attaching an SSM role to it, figuring out it's platform type and patching it accordingly",
        schemaVersion: '0.3',
        assumeRole,
        parameters: {
            EC2IamRole: {
                type: 'String',
                description: 'IAM Instance Profile Name, usually is the name of the IAM role',
                default: 'MyEC2SSMRole',
            },
            Application: {
                type: 'String',
                description: 'Choose from one of the choices below to install',
                allowedValues: ['mariadb-server', 'httpd'],
            },
        },
        mainSteps: [
            {
                name: 'wait_for_ssm_info',
                action: 'aws:waitForAwsResourceProperty',
                inputs: {
                    Service: 'ssm',
                    Api: 'DescribeInstanceInformation',
                    Filters: [{ Key: 'InstanceIds', Values: ['{{create_ec2.InstanceIds}}'] }],
                    PropertySelector: '$.InstanceInformationList[0].PingStatus',
                    DesiredValues: ['Online'],
                },
            },
            {
                name: 'DeployApplicationByChoice',
                action: 'aws:branch',
                inputs: {
                    Choices: [
                        {
                            NextStep: 'InstallConfigureMariaDB',
                            Variable: '{{Application}}',
                            StringEquals: 'mariadb-server',
                        },
                        {
                            NextStep: 'InstallConfigureHttpd',
                            Variable: '{{Application}}',
                            StringEquals: 'httpd',
                        },
                    ],
                    Default: 'InstallConfigureHttpd',
                },
            },
            {
                name: 'InstallConfigureMariaDB',
                action: 'aws:runCommand',
                inputs: {
                    DocumentName: 'AWS-RunShellScript',
                    InstanceIds: ['{{create_ec2.InstanceIds}}'],
                    Parameters: {
                        commands: [
                            'yum -y install {{Application}}',
                            'systemctl enable mariadb; systemctl start mariadb',
                            'sleep 5',
                            'aws configure set default.region us-east-1',
                            "mysqladmin password $(aws ssm get-parameter --name mysql-pass --with-decryption --output text --query 'Parameter.Value')",
                        ],
                    },
                },
                isEnd: true,
            },
            {
                name: 'InstallConfigureHttpd',
                action: 'aws:runCommand',
                inputs: {
                    DocumentName: 'AWS-RunShellScript',
                    InstanceIds: ['{{create_ec2.InstanceIds}}'],
                    Parameters: {
                        commands: [
                            'yum -y install {{Application}}',
                            'systemctl enable {{Application}}; systemctl start {{Application}}',
                        ],
                    },
                },
                isEnd: true,
            },
        ],
    };
};
