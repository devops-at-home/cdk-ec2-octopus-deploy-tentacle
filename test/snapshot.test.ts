import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { OctopusDeployDocument } from '../src/index';

describe('Snapshot test', () => {
    const stack = new Stack();

    new OctopusDeployDocument(stack, 'Ec2OctopusDeploy', {
        secretArn: 'secretArn',
    });

    const template = Template.fromStack(stack);

    test('Snapshot test', () => {
        expect(template.toJSON()).toMatchSnapshot();
    });
});
