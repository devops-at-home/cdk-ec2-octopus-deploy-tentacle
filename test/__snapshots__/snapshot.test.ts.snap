// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test Snapshot test 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Ec2OctopusDeployCfnDocument920D1912": Object {
      "Properties": Object {
        "Content": Object {
          "assumeRole": Object {
            "Ref": "Ec2OctopusDeployRole3425770F",
          },
          "description": "Automation for creating an instance, attaching an SSM role to it, figuring out it's platform type and patching it accordingly",
          "mainSteps": Array [
            Object {
              "action": "aws:waitForAwsResourceProperty",
              "inputs": Object {
                "Api": "DescribeInstanceInformation",
                "DesiredValues": Array [
                  "Online",
                ],
                "Filters": Array [
                  Object {
                    "Key": "InstanceIds",
                    "Values": Array [
                      "{{create_ec2.InstanceIds}}",
                    ],
                  },
                ],
                "PropertySelector": "$.InstanceInformationList[0].PingStatus",
                "Service": "ssm",
              },
              "name": "wait_for_ssm_info",
            },
            Object {
              "action": "aws:branch",
              "inputs": Object {
                "Choices": Array [
                  Object {
                    "NextStep": "InstallConfigureMariaDB",
                    "StringEquals": "mariadb-server",
                    "Variable": "{{Application}}",
                  },
                  Object {
                    "NextStep": "InstallConfigureHttpd",
                    "StringEquals": "httpd",
                    "Variable": "{{Application}}",
                  },
                ],
                "Default": "InstallConfigureHttpd",
              },
              "name": "DeployApplicationByChoice",
            },
            Object {
              "action": "aws:runCommand",
              "inputs": Object {
                "DocumentName": "AWS-RunShellScript",
                "InstanceIds": Array [
                  "{{create_ec2.InstanceIds}}",
                ],
                "Parameters": Object {
                  "commands": Array [
                    "yum -y install {{Application}}",
                    "systemctl enable mariadb; systemctl start mariadb",
                    "sleep 5",
                    "aws configure set default.region us-east-1",
                    "mysqladmin password $(aws ssm get-parameter --name mysql-pass --with-decryption --output text --query 'Parameter.Value')",
                  ],
                },
              },
              "isEnd": true,
              "name": "InstallConfigureMariaDB",
            },
            Object {
              "action": "aws:runCommand",
              "inputs": Object {
                "DocumentName": "AWS-RunShellScript",
                "InstanceIds": Array [
                  "{{create_ec2.InstanceIds}}",
                ],
                "Parameters": Object {
                  "commands": Array [
                    "yum -y install {{Application}}",
                    "systemctl enable {{Application}}; systemctl start {{Application}}",
                  ],
                },
              },
              "isEnd": true,
              "name": "InstallConfigureHttpd",
            },
          ],
          "parameters": Object {
            "Application": Object {
              "allowedValues": Array [
                "mariadb-server",
                "httpd",
              ],
              "description": "Choose from one of the choices below to install",
              "type": "String",
            },
            "EC2IamRole": Object {
              "default": "MyEC2SSMRole",
              "description": "IAM Instance Profile Name, usually is the name of the IAM role",
              "type": "String",
            },
          },
          "schemaVersion": "0.3",
        },
        "DocumentType": "Automation",
        "Name": "ConfigureOctopusDeploy",
        "TargetType": "/AWS::EC2::Instance",
      },
      "Type": "AWS::SSM::Document",
    },
    "Ec2OctopusDeployRole3425770F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ssm.amazon.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
